src/memory_manager.rs

```diff
@@ -853,4 +1240,188 @@ mod test {
+        assert_eq!(memory_2.grow(1), 0);
+
+        memory_0.write(0, &[1, 2, 3]);
+        memory_1.write(0, &[4, 5, 6]);
+        memory_2.write(0, &[7, 8, 9]);
+
+        let mut mem_mgr = MemoryManager::init(mem);
+        let memory_0 = mem_mgr.get(MemoryId(0));
+        let memory_1 = mem_mgr.get(MemoryId(1));
+        let memory_2 = mem_mgr.get(MemoryId(2));
+
+        let mut bytes = vec![0; 3];
+        // Check that data is correctly reinitialized.
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, vec![1, 2, 3]);
+
+        memory_1.read(0, &mut bytes);
+        assert_eq!(bytes, vec![4, 5, 6]);
+
+        memory_2.read(0, &mut bytes);
+        assert_eq!(bytes, vec![7, 8, 9]);
+
+        // Free MemoryId 1.
+        mem_mgr.free(MemoryId(1));
+
+        // Check that data of MemoryId 0 and 2 is correctly reinitialized.
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, vec![1, 2, 3]);
+
+        memory_2.read(0, &mut bytes);
+        assert_eq!(bytes, vec![7, 8, 9]);
+
+        // Check that date of MemoryId 1 is freed.
+        assert_eq!(memory_1.size(), 0);
+        memory_1.read(0, &mut bytes);
+    }
+
+    #[test]
+    fn freeing_already_free_memory() {
+        let mut mem_mgr = MemoryManager::init(make_memory());
+        let memory_0 = mem_mgr.get(MemoryId(0));
+
+        assert_eq!(memory_0.grow(1), 0);
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 1);
+
+        mem_mgr.free(MemoryId(0));
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 0);
+
+        mem_mgr.free(MemoryId(0));
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 0);
+    }
+
+    #[test]
+    fn grow_memory_after_freeing_it() {
+        let mut mem_mgr = MemoryManager::init(make_memory());
+        let memory_0 = mem_mgr.get(MemoryId(0));
+
+        // grow and write to memory
+        assert_eq!(memory_0.grow(1), 0);
+        memory_0.write(0, &[7, 1, 5]);
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 1);
+
+        let mut bytes = vec![0; 3];
+
+        // read from memory
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, &[7, 1, 5]);
+
+        // free memory
+        mem_mgr.free(MemoryId(0));
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 0);
+
+        // grow memory
+        assert_eq!(memory_0.grow(1), 0);
+
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 1);
+
+        // check that old bucket is reassign to the memory
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, &[7, 1, 5]);
+
+        // try growing once more
+        assert_eq!(memory_0.grow(1), 1);
+        assert_eq!(mem_mgr.get(MemoryId(0)).size(), 2);
+    }
+
+    #[test]
+    fn free_memory_that_was_not_used() {
+        let mut mem_mgr = MemoryManager::init(make_memory());
+        let memory_0 = mem_mgr.get(MemoryId(0));
+        assert_eq!(memory_0.grow(1), 0);
+
+        mem_mgr.free(MemoryId(5));
+    }
+
+    #[test]
+    fn upgrade_from_v1_to_v2() {
```

Comment:
Is this enough? Upgrading from V1 to V2, if it has a bug, can lead to data loss, which is really bad. This test is testing a very simple use-case. We should ideally have a `proptest`, where we initialize a V1 memory manager with a random number of memories, write random data, then reload with v2, and verify that all the memories and the data exist. Have a look at `write_and_read_random_bytes` for inspiration.